### PR TITLE
[codex] Complete Milestone 2 workflow and artifact safety

### LINKED ISSUE #16: M2-05: First-Class BDD Stage
# M2-05 - First-Class BDD Stage

## Goal
Move BDD generation into `ec bdd` and remove `.feature` generation from `ideate()`.

## Scope
- Add `bdd` workflow stage
- Add `ec bdd <feature-id>`
- Add orchestrator `bdd()` method
- Generate `.feature` files from stories and acceptance criteria
- Remove placeholder BDD creation from `ideate`

## Expected Input
```bash
ec bdd <feature-id>
```

## Reads
- `.ec/features/<feature-id>/feature.yaml`
- `.ec/features/<feature-id>/plan.md`
- `.ec/features/<feature-id>/stories/*.md`

## Writes
- `.ec/features/<feature-id>/bdd/*.feature`
- `.ec/runs/<run-id>.json`

## Acceptance Criteria
- `ec bdd <feature-id>` generates `.feature` files
- `ideate` no longer writes `.feature` files
- The same input produces stable output
- Missing stories produce a clear failure message

## Testing
- AC-backed tests for BDD generation
- Structural tests for stage transition
- Regression test confirming `ideate` does not write BDD

## Output Contract Requirement
The command output must include:
- command name
- feature id where applicable
- input summary
- files read
- files written
- result status
- next recommended command

### LINKED ISSUE #17: M2-04: Structured Stories and Acceptance Criteria
# M2-04 - Structured Stories and Acceptance Criteria

## Goal
Make `plan` generate stable story and acceptance-criteria artifacts for downstream stages.

## Scope
- Add `Story` type
- Add `AcceptanceCriteria` type
- Update `plan` output
- Normalize story markdown format
- Detect missing or malformed acceptance criteria

## Expected Input
```bash
ec plan <feature-id>
```

## Reads
- `.ec/features/<feature-id>/feature.yaml`
- `.ec/features/<feature-id>/ideation.md`

## Writes
- `.ec/features/<feature-id>/plan.md`
- `.ec/features/<feature-id>/stories/*.md`
- `.ec/features/<feature-id>/implementation/tasks.md`
- `.ec/runs/<run-id>.json`

## Acceptance Criteria
- Each story has a user story block
- Each story has explicit numbered acceptance criteria
- Acceptance criteria are stable enough to map to BDD
- Malformed story or acceptance-criteria output is detectable

## Testing
- AC-backed tests for expected story structure
- Structural tests for parser and formatter
- Edge-case tests for missing acceptance criteria

## Output Contract Requirement
The command output must include:
- command name
- feature id where applicable
- input summary
- files read
- files written
- result status
- next recommended command

### LINKED ISSUE #18: M2-06: Traceability Map
# M2-06 - Traceability Map

## Goal
Generate and validate `traceability/traceability-map.json`.

## Scope
- Define a traceability schema
- Generate `traceability-map.json`
- Link stories to acceptance criteria
- Link acceptance criteria to BDD
- Link stories to implementation tasks where possible
- Detect broken or missing links

## Expected Input
Usually generated during or after:

```bash
ec bdd <feature-id>
```

## Reads
- `.ec/features/<feature-id>/plan.md`
- `.ec/features/<feature-id>/stories/*.md`
- `.ec/features/<feature-id>/bdd/*.feature`
- `.ec/features/<feature-id>/implementation/tasks.md`

## Writes
- `.ec/features/<feature-id>/traceability/traceability-map.json`

## Acceptance Criteria
- Every acceptance criterion maps to at least one BDD scenario
- Orphaned stories, acceptance criteria, and scenarios are detectable
- Generated map is deterministic

## Testing
- Structural tests for graph generation
- Tests for missing references
- Tests for deterministic output

## Output Contract Requirement
The command output must include:
- command name
- feature id where applicable
- input summary
- files read
- files written
- result status
- next recommended command

### LINKED ISSUE #19: M2-07: Consistency Review Stage
# M2-07 - Consistency Review Stage

## Goal
Add `ec review-consistency` to detect lifecycle inconsistencies before implementation.

## Scope
- Add `review-consistency` workflow stage
- Add CLI command
- Generate `review/consistency-review.md`
- Check missing acceptance criteria, BDD coverage, ambiguity, non-testable acceptance criteria, and artifact drift

## Expected Input
```bash
ec review-consistency <feature-id>
```

## Reads
- `.ec/features/<feature-id>/feature.yaml`
- `.ec/features/<feature-id>/plan.md`
- `.ec/features/<feature-id>/stories/*.md`
- `.ec/features/<feature-id>/bdd/*.feature`
- `.ec/features/<feature-id>/traceability/traceability-map.json`

## Writes
- `.ec/features/<feature-id>/review/consistency-review.md`
- `.ec/runs/<run-id>.json`

## Acceptance Criteria
- Missing acceptance-criteria coverage is detected
- Missing BDD mappings are detected
- Non-testable acceptance criteria are flagged
- Artifact drift is reported

## Testing
- AC-backed tests for expected detection behavior
- Structural tests for partial or incomplete features
- Output tests for review report format

## Output Contract Requirement
The command output must include:
- command name
- feature id where applicable
- input summary
- files read
- files written
- result status
- next recommended command

### LINKED ISSUE #20: M2-09: Cleanup and Workflow Enforcement
# M2-09 - Cleanup and Workflow Enforcement

## Goal
Remove placeholder logic, enforce workflow integrity, and add regression and end-to-end coverage.

## Scope
- Remove obsolete placeholder logic
- Enforce valid stage transitions
- Enforce required artifact presence
- Add regression tests
- Add an end-to-end workflow test

## Expected Input
Representative workflow:

```bash
ec ideate "Add first-class BDD workflow stage"
ec plan <feature-id>
ec bdd <feature-id>
ec review-consistency <feature-id>
```

## Reads
- Full feature workspace state
- `.ec/config`
- Run history as needed

## Writes
- No new feature-behavior artifacts beyond normal workflow outputs

## Acceptance Criteria
- No placeholder BDD generation remains in `ideate`
- Invalid stage transitions fail clearly
- Missing required artifacts fail clearly
- Full Milestone 2 workflow passes end to end

## Testing
- Regression tests for existing commands
- End-to-end test for `ideate -> plan -> bdd -> review-consistency`
- No new feature-level unit tests unless tied to existing acceptance criteria

## Output Contract Requirement
The command output must include:
- command name
- feature id where applicable
- input summary
- files read
- files written
- result status
- next recommended command

### PR DESCRIPTION
**Pull request by apkasten906**
## What changed
This PR completes the Milestone 2 workflow expansion for Ethereal-CLAW.

It adds first-class `bdd` and `review-consistency` stages, upgrades the CLI output contract, strengthens feature status/reporting, and formalizes artifact synchronization and conflict handling across the workflow.

Key changes:
- add `bdd` and `review-consistency` commands and stage wiring
- move story generation fully into `plan`
- generate and validate `traceability/traceability-map.json`
- add structured story parsing/rendering and consistency review validation
- add human-readable command reports with `--json` automation support
- add safer duplicate-feature handling for `ideate`, including explicit overwrite, cancellation, and non-interactive conflict behavior
- make workspace overwrite transactional with staged replacement and warning-only post-commit cleanup failure
- enforce divergence protection across regenerated stage artifacts instead of silently overwriting user-edited files
- document the write-safety and conflict-handling policy in the blueprint and user-facing docs

## Why it changed
Milestone 2 needed the workflow to move from a loose scaffold into a stateful, auditable SDLC manager with explicit stage boundaries, synchronization checks, and automation-safe behavior.

The branch also addresses safety issues found during review:
- repeated runs needed to be deterministic where artifacts are synchronization-sensitive
- duplicate feature creation needed explicit overwrite semantics
- automation mode needed deterministic conflict errors instead of interactive prompts
- regenerated artifacts needed fail-closed behavior instead of silent replacement

## User and developer impact
Users now get:
- a complete `ideate -> plan -> bdd -> review-consistency -> implement -> test -> review` flow
- clearer command output by default
- safer overwrite/conflict behavior for feature workspaces
- explicit artifact drift detection before later stages proceed

Developers now get:
- stronger workflow/state tests
- explicit artifact ownership and synchronization rules
- documented write-safety policy for future stage implementations

## Validation
Ran:
- `npm run typecheck`
- `npm run lint`
- `npm test`

Current result: full test suite passes (`66` tests).

### CODE REVIEW

File: packages/cli/src/presentation/command-report.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    "Written",
+    ...formatItems(report.written),
+    "",
+    "Budget",
+    ...(report.budget
+      ? [
+        `- Provider: ${report.provider ?? "unknown"}`,
+        `- Model tier: ${report.budget.modelTier}`,
+        `- Estimated tokens/cost: ${report.budget.estimatedInputTokens} input / ${report.budget.estimatedOutputTokens} output, $${report.budget.estimatedCostUsd.toFixed(4)}`,
+        `- Actual tokens/cost: ${report.budget.actualPromptTokens} input / ${report.budget.actualCompletionTokens} output, $${report.budget.estimatedCostUsd.toFixed(4)}`
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/cli/src/presentation/command-report.ts]: Updated locally. The human-readable report now says `Actual tokens (estimated cost)` to reflect that only token counts are actual while the cost remains derived from `estimatedCostUsd`. I added a smoke-test assertion for the revised label.

File: packages/core/src/artifacts/feature-structure-service.ts [hunk 2]
```diff
@@ -24,34 +37,29 @@ export class FeatureStructureService {
...
-        title: feature.title,
-        request: feature.request,
-        status: feature.status,
-        createdAt: feature.createdAt,
-        updatedAt: feature.updatedAt
-      }, { lineWidth: -1 })
-    );
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/core/src/artifacts/feature-structure-service.ts:51]: Updated locally. `workspaceExists()` now checks the feature root instead of only `feature.yaml`, so an existing partial/corrupt feature directory is treated as a conflict and ideation fails closed unless overwrite is explicit. I also added regression coverage for the directory-without-metadata case in `tests/unit/feature-structure-service.test.ts`.

File: docs/workflow.md [hunk 3]
```diff
@@ -8,28 +8,58 @@ For exact syntax and options, see the [command reference](command-reference.md).
 
 | Stage | Command | Input | Main outputs | Prompt docs |
 | --- | --- | --- | --- | --- |
-| Ideate | `ethereal ideate "<request>"` | Rough feature request | Feature workspace, metadata, ideation, initial story, BDD placeholder | [Ideation](prompts/ideation.md), [Story writer](prompts/story-writer.md) |
-| Plan | `ethereal plan <feature-slug>` | Existing feature workspace | `plan.md`, `implementation/tasks.md` | [Planner](prompts/planner.md) |
+| Ideate | `ethereal ideate "<request>"` | Rough feature request | Feature workspace, metadata, ideation | [Ideation](prompts/ideation.md) |
+| Plan | `ethereal plan <feature-slug>` | Existing feature workspace | `plan.md`, `stories/*.md`, `implementation/tasks.md` | [Planner](prompts/planner.md), [Story writer](prompts/story-writer.md) |
+| BDD | `ethereal bdd <feature-slug>` | Planned feature workspace | `bdd/*.feature`, `traceability/traceability-map.json` | [Story writer](prompts/story-writer.md) |
+| Review Consistency | `ethereal review-consistency <feature-slug>` | Planned + BDD feature workspace | `review/consistency-review.md` | [Reviewer](prompts/reviewer.md) |
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply docs/workflow.md]: Updated locally. The workflow table now marks `bdd` as `n/a (local transformation)` and `review-consistency` as `n/a (local validation)` so the docs no longer imply provider execution for those stages.

File: packages/core/src/orchestration/workflow-orchestrator.ts [hunk 4]
```diff
@@ -83,15 +121,111 @@ export class WorkflowOrchestrator {
...
+      await this.ensureFeatureArtifacts(
+        featureSlug,
+        ["plan.md", path.join("stories", "001-initial-story.md"), path.join("bdd", "001-initial.feature"), path.join("traceability", "traceability-map.json")],
+        "Consistency review requires planned stories, BDD, and traceability artifacts. Run `ec plan` and `ec bdd` first."
+      );
+      const review = await this.buildConsistencyReview(featureSlug);
+      const reviewPath = path.join("review", "consistency-review.md");
+      await this.assertSafeToWriteArtifact(featureSlug, reviewPath, review);
+      await this.artifacts.writeFeatureArtifact(featureSlug, reviewPath, review);
+
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/core/src/orchestration/workflow-orchestrator.ts:227]: Updated in `cb7014c`. `buildConsistencyReview()` now returns a structured result with `content`, `status`, and `findings`, and `reviewConsistency()` passes `success=false` to `finishRun(...)` whenever the report status is `needs-attention`. That keeps the run log and feature status fail-closed instead of advancing to `consistency-reviewed` on a failing consistency gate. I also added regression coverage asserting `review-consistency` returns `success=false` when drift is detected.

File: tests/unit/workflow-orchestrator.test.ts [hunk 5]
```diff
@@ -82,8 +87,17 @@ describe("WorkflowOrchestrator", () => {
...
+      "plan",
+      "bdd",
+      "review-consistency",
+      "implement",
+      "test",
+      "review"
+    ]);
+    expect(beforeRun).toContain("status: ideated");
+    expect(afterRun).toContain("status: reviewed");
+    expect(afterRun).toContain("request: refresh tokens for admins");
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply tests/unit/workflow-orchestrator.test.ts:100]: Updated in `cb7014c`. I renamed the test to `does not rerun ideation or rewrite the saved feature request during a full run`, which matches the current assertions: the status still advances through the workflow, but the saved request remains stable across `run()`.

File: packages/cli/src/commands/command-support.ts [hunk 6]
```diff
@@ -33,3 +40,220 @@ export async function resolveStageRequest(featureSlug: string, requestOverride:
...
+        written: [
+          `${featureRoot}/tests/test-plan.md`,
+          `${featureRoot}/tests/generated-tests.md`,
+          path.relative(rootDir, workspace.runsDirectory).replaceAll("\\", "/")
+        ],
+        next: `ec review ${featureSlug}`
+      };
+    case "review":
+      return {
+        read: [configPath, `${featureRoot}/stories/001-initial-story.md`, `${featureRoot}/tests/test-plan.md`, `${featureRoot}/tests/generated-tests.md`],
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/cli/src/commands/command-support.ts]: Updated in `cb7014c`. I made the same alignment change for `review`: its command report now only lists the project config under `Read`, matching the current implementation instead of implying that the story/test artifacts are directly read by the stage.

File: packages/cli/src/commands/command-support.ts [hunk 7]
```diff
@@ -33,3 +40,220 @@ export async function resolveStageRequest(featureSlug: string, requestOverride:
...
+        read: [configPath, `${featureRoot}/review/consistency-review.md`],
+        written: [
+          `${featureRoot}/implementation/change-summary.md`,
+          path.relative(rootDir, workspace.runsDirectory).replaceAll("\\", "/")
+        ],
+        next: `ec test ${featureSlug}`
+      };
+    case "test":
+      return {
+        read: [configPath, `${featureRoot}/implementation/change-summary.md`, `${featureRoot}/stories/001-initial-story.md`, `${featureRoot}/bdd/001-initial.feature`],
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/cli/src/commands/command-support.ts]: Updated in `cb7014c`. I trimmed the `test` command report `Read` list down to the files the stage actually reads today: the project config and `implementation/change-summary.md`. That keeps the human-readable audit output aligned with real IO instead of implied context.

File: packages/core/src/orchestration/workflow-orchestrator.ts [hunk 8]
```diff
@@ -105,10 +239,16 @@ export class WorkflowOrchestrator {
     await this.ensureExistingFeatureWorkspace(featureSlug);
 
     try {
+      await this.ensureFeatureArtifacts(
+        featureSlug,
+        [path.join("review", "consistency-review.md")],
+        "Implementation planning requires a completed consistency review. Run `ec review-consistency` first."
+      );
```
[the remark this answers is not in the document]
Comment (apkasten906) [reply packages/core/src/orchestration/workflow-orchestrator.ts:257]: Updated in `cb7014c`. `implement()` now parses `review/consistency-review.md` via `ensureConsistencyReviewPassed()` and refuses to proceed unless the artifact contains `Status: ready-for-implement`. If the report says `needs-attention` or the status line is malformed, the command now fails closed with an explicit rerun message. I added a regression test covering the `needs-attention` case.
